documentation/projects/proposals/monitoring/20230606_implementation_plan_ecs_alarms.md

```diff
@@ -0,0 +1,595 @@
+
+```{note}
+The CloudWatch Dashboards that collect all metrics for `next` root module services
+will not be moved into these new modules. The dashboards include all services defined
+in the root module and keeping it in the root module level allows us to easily share
+documentation relevant for all services and simplify the configuration as proposed
+in [WordPress/openverse-infrastructure #472](https://github.com/WordPress/openverse-infrastructure/issues/472).
+```
+
+We will need at least one new SNS topic for unstable alarms to send
+notifications to a channel specific for unstable alarms. Separate notification
+channels for unstable and stable alerts helps maintainers quickly distinguish
+between the types of alerts they respond to. The new topic should be created in
+the `singleton` module and follow the existing pattern of deferring management
+to the production environment by using a data resource in staging. The existing
+topics and subscription should also be moved into the singleton module.
+
+### Stable vs Unstable Alarms and their Responses
+
+As this project will add several new alarms, it is critical to understand the
+distinction between stable and unstable alarms. In particular, all maintainers
+interacting with the new tools must understand how to respond and determine
+urgency for all alerts. Stability also refers to individual alarms and must be
+considered in context of how they relate to the alerts that depend on them.
+Alarms that comprise a composite alarm, for example, may require a different
+approach to stability and accuracy if the resulting composite alarm is itself
+sufficiently accurate.
+
+Alarms should be considered unstable if they regularly produce false positives.
+That is to say, if we incorrectly send an alert, and this happens on a regular
+basis, then the alarm is unstable. Stable alarms, on the other hand, should have
+virtually no incidents of false positives. If an alarm sends alerts to the
+stable notification channel, it should be assumed that something is wrong that
+needs to be addressed. In either case, the
+[alert's corresponding run book](#run-books) should guide maintainers in
+identifying severity so that they can complete the rest of the triaging process
+outlined in the general incident response guide. It is important to note that
+unstable alarms may produce a high number of false positives but should not be
+treated with the blanket assumption that they are entirely inaccurate. As
+described below, **run books must assist maintainers in quickly identifying
+false positives**.
+
+Because all the alarms proposed in this plan are new and for services that have
+never had alarms of this nature, all of them will begin as unstable alarms. For
+each alarm, until it is deemed stable (after we have observed for a sufficient
+period of time that it does not create false positives), alarms will follow the
+process outlined below for "unstable alarm response".
+
+### Stabilising Alarms
```

Comment:
Do you think it may also make sense to have a DAG or other process periodically check that an alarm hasn't been "unstable" for too long a period (say, 6 months)? Just to ensure that we don't have any alarms that fall through the cracks or remain unstable indefinitely if something happens to the assigned maintainer?

Reply:
Not a bad idea and probably low effort to implement. I need to check if AWS exposes such information and will add a step for the DAG if so.

Reply:
AWS only provides alarm history for 14 days. To track how long it's been since an alarm last alarmed we'd need to create a separate action on relevant alarms that stores that data somewhere. I don't think that's feasible in this IP and would significantly increase the scope. If in the future we find that an alarm has been "too quite" for too long and this cause us to miss an outage, we can invest the time and infrastructure into this then.